### PR TITLE
feat(subs): redis backend

### PR DESCRIPTION
**Pull request by michael-0acf4**
* Redis Backend base logic port + some improvements
* Moved `SUBSTANTIAL_POLL_INTERVAL_SEC` and `SUBSTANTIAL_LEASE_LIFESPAN_SEC` to config

#### Migration notes
* Renamed `Backend.fs()` and `Backend.memory()` to `Backend.dev_fs()` and `Backend.dev_memory()` 
* Removed `SUBSTANTIAL_RELAUNCH_MS` as it was relevant only for purely worker-based runs, which rendered the new `SUBSTANTIAL_POLL_INTERVAL_SEC` redundant when an interrupt hits.

- [x] The change comes with new or modified tests
- [ ] Hard-to-understand functions have explanatory comments
- [x] End-user documentation is updated to reflect the change

## Release Notes

- **New Features**
	- Introduced Redis as a new backend option for enhanced data management.
	- Added Docker Compose configuration for a Redis service.
	- Implemented comprehensive testing for Redis functionality and backend integration.

- **Bug Fixes**
	- Improved error handling during backend initialization.

- **Documentation**
	- Updated type definitions for backend configurations to streamline Redis integration.

- **Chores**
	- Refactored test cases for clarity and consistency across different backend types.

### CODE REVIEW

**Comment by michael-0acf4** [comment]
@coderabbitai review

File: src/substantial/tests/mod.rs [hunk 1]
```diff
@@ -73,22 +74,47 @@ mod tests {
...
-            }),
+    fn test_state_consistency_logic() {
+        let backends: Vec<(&str, Box<dyn Backend>)> = vec![
+            ("memory", Box::new(MemoryBackend::default().unwrap())),
+            (
+                "fs",
+                Box::new({
+                    let root = PathBuf::from("tmp/test_two/substantial");
+                    let backend = FsBackend::new(root.clone()).unwrap();
+                    std::fs::remove_dir_all(root).ok();
```
[the remark this answers is not in the document]
Comment (michael-0acf4) [reply src/substantial/tests/mod.rs:85]: The root and prefix will account for that

File: src/typegate/src/runtimes/substantial.ts [hunk 2]
```diff
@@ -201,14 +211,14 @@ export class SubstantialRuntime extends Runtime {
...
-      const completed = [];
+      const ongoing = [] as Array<QueryOngoingWorkflowResult>;
+      const completed = [] as Array<QueryCompletedWorkflowResult>;
       for (const runId of relatedRuns) {
         const run = await this.agent.retrieveEvents(runId);
         const startedAt = run.operations[0]?.at;
         if (!startedAt) continue;
 
-        let endedAt, result: any;
+        let endedAt: string, result: any;
```
[the remark this answers is not in the document]
Comment (michael-0acf4) [reply src/typegate/src/runtimes/substantial.ts:221]: Type shape is only known by the typegraph author

File: src/substantial/src/backends/redis.rs [hunk 3]
```diff
@@ -0,0 +1,510 @@
...
+                let exp_time = DateTime::parse_from_rfc3339(&exp_time)?.to_utc();
+                if exp_time > Utc::now() {
+                    let run_id =
+                        self.parts(&sched_ref)?.last().cloned().with_context(|| {
+                            format!("Could not parse run_id from {:?}", sched_ref)
+                        })?;
+
+                    active_lease_ids.push(run_id);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (michael-0acf4) [reply src/substantial/src/backends/redis.rs:310]: No need, it must fail fast.

File: src/substantial/src/backends/redis.rs [hunk 4]
```diff
@@ -0,0 +1,510 @@
...
+            let raw_logs: Vec<Vec<u8>> = script.key(log_key).arg(run_id).invoke(r)?;
+
+            let mut ret = Vec::new();
+            for raw_log in raw_logs.iter() {
+                ret.push(Metadata::parse_from_bytes(raw_log)?);
+            }
+
+            Ok(ret)
+        })
+    }
```
[the remark this answers is not in the document]
Comment (michael-0acf4) [reply src/substantial/src/backends/redis.rs:454]: That's interesting, but I cannot see any alternatives other than using ` redis.call("SCAN", 0, "MATCH", prefix .. "*")` which will traverse ALL keys, hence storing a reference in a list (which I can get through a single call) and using that reference to point at the actual content. wdyt?

File: src/typegate/engine/src/runtimes/substantial.rs [hunk 5]
```diff
@@ -22,18 +22,21 @@ pub struct Ctx {
...
-            Rc::new(FsBackend::new(root).unwrap())
+            Ok(Rc::new(FsBackend::new(root).unwrap()))
         }
-        SubstantialBackend::Memory => Rc::new(MemoryBackend::default().unwrap()),
-        SubstantialBackend::Redis(_) => todo!(),
+        SubstantialBackend::Memory => Ok(Rc::new(MemoryBackend::default().unwrap())),
+        SubstantialBackend::Redis(cfg) => Ok(Rc::new(RedisBackend::new(
+            cfg.connection_string.clone(),
+            Some("typegate".to_owned()),
+        )?)),
```
[the remark this answers is not in the document]
Comment (michael-0acf4) [reply src/typegate/engine/src/runtimes/substantial.rs:39]: `.unwrap()` returns the wrapped `KeyValueBackend` type and does not came from a Result or Optional in this case.

File: tools/tasks/install.ts [hunk 6]
```diff
@@ -85,27 +85,33 @@ export default {
...
         },
       });
     },
   },
+  // "install-protobuf-compiler": {
+  //   fn: ($) => $`sudo apt install protobuf-compiler`,
+  // },
+  // Enable rust codegen for protoc
+  "install-protobuf-rust": {
+    fn: ($) => $`cargo install protobuf-codegen@3.5.1`,
```
Comment (Yohe-Am) [inline tools/tasks/install.ts:115]: You can use cargo binstall for this. Or better yet, use a `cargobi` port.

File: tests/runtimes/substantial/common.ts [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+              {
+                result: {
+                  status: "COMPLETED",
+                  value: 'Email sent to one@example.com: "confirmed!"',
+                },
+                run_id: runIds[0],
+              },
+              {
+                result: {
+                  status: "COMPLETED_WITH_ERROR",
```
Comment (Yohe-Am) [inline tests/runtimes/substantial/common.ts:298]: Are retry policies and stuff in scope for substantial?

**Review by Yohe-Am — APPROVED** [review]
Nice work.

File: tests/runtimes/substantial/common.ts (hunk 7 above)
Comment (michael-0acf4) [reply tests/runtimes/substantial/common.ts:298]: Yes, something like `ctx.save(() => ..., { backoff: Time.seconds(10), retries: 10 })`.

File: tools/tasks/install.ts (hunk 6 above)
Comment (michael-0acf4) [reply tools/tasks/install.ts:115]: Thanks, I'll keep this in mind when tackling other stuff.

**Pull request merged**